src/cargo/sources/registry/remote.rs

```diff
@@ -13,16 +15,23 @@ use ops;
 use sources::git;
 use sources::registry::{RegistryData, RegistryConfig, INDEX_LOCK};
 use util::network;
-use util::paths;
-use util::{FileLock, Filesystem};
+use util::{FileLock, Filesystem, LazyCell};
 use util::{Config, CargoResult, ChainError, human, Sha256, ToUrl};
 
 pub struct RemoteRegistry<'cfg> {
     index_path: Filesystem,
     cache_path: Filesystem,
     source_id: SourceId,
     config: &'cfg Config,
-    handle: Option<Easy>,
+    handle: LazyCell<RefCell<Easy>>,
+
+    // Note that this `'static lifetime here is actually a lie, it's actually a
+    // borrow into the `repo` object below. We're guaranteed, though, that if
+    // filled in `tree` will be destroyed first, so this should be ok.
```

Comment:
Oh sorry missed this! Yes the unspecified drop order is what guarantees this. Lots of projects are relying on this so I don't think it's necessary to go out of the way and use `ManuallyDrop`, and yeah I'd also prefer to keep Cargo on stable.

@matklad unfortunately that crate won't help as it's targeted at Rust pointers, whereas here it's all phantom lifetimes through libgit2 :(